hack/generator/pkg/config/configuration.go

```diff
@@ -184,24 +184,77 @@ func (config *Configuration) initialize(configPath string) error {
 	return kerrors.NewAggregate(errs)
 }
 
-// ShouldExport tests for whether a given type should be exported as Go code
-// Returns a result indicating whether export should occur as well as a reason for logging
-func (config *Configuration) ShouldExport(typeName astmodel.TypeName) (result ShouldExportResult, because string) {
-	for _, f := range config.ExportFilters {
-		if f.AppliesToType(typeName) {
-			switch f.Action {
-			case ExportFilterExclude:
+type ExportFilterFunc func(astmodel.TypeName) (result ShouldExportResult, because string)
+
+func buildExportFilterFunc(f *ExportFilter, allTypes astmodel.Types) ExportFilterFunc {
+	switch f.Action {
+	case ExportFilterExclude:
+		return func(typeName astmodel.TypeName) (ShouldExportResult, string) {
+			if f.AppliesToType(typeName) {
 				return Skip, f.Because
-			case ExportFilterInclude:
+			}
+
+			return "", ""
+		}
+
+	case ExportFilterInclude:
+		return func(typeName astmodel.TypeName) (ShouldExportResult, string) {
+			if f.AppliesToType(typeName) {
 				return Export, f.Because
-			default:
-				panic(errors.Errorf("unknown exportfilter directive: %s", f.Action))
+			}
+
+			return "", ""
+		}
+
+	case ExportFilterIncludeTransitive:
+		applicableTypes := make(astmodel.TypeNameSet)
+		for tn := range allTypes {
+			if f.AppliesToType(tn) {
+				getAllReferencedTypes(allTypes, tn, applicableTypes)
 			}
 		}
+
+		return func(typeName astmodel.TypeName) (ShouldExportResult, string) {
+			if applicableTypes.Contains(typeName) {
+				return Export, f.Because
+			}
+
+			return "", ""
+		}
+
+	default:
+		panic(errors.Errorf("unknown exportfilter directive: %s", f.Action))
+	}
+}
+
+func getAllReferencedTypes(allTypes astmodel.Types, root astmodel.TypeName, output astmodel.TypeNameSet) {
```

Comment:
You possibly could make this a type w/ `allTypes` and `output` as struct variables...

```
type TransitiveTypeFinder struct {
    allTypes astmodel.Types
    applicableTypes astmodel.TypeNameSet
    filter *ExportFilter
}

func (tf *TransitiveTypeFinder) ShouldExport(typeName astmodel.TypeName) (ShouldExportResult, string) {
    if tf.applicableTypes.Contains(typeName) {
        return Export, tf.filter.Because
    }
    return "", ""
}

func (tf *TransitiveTypeFinder) Initialize() { // Or whatever name is good
    for tn := range tf.allTypes {
        if tf.filter.AppliestoType(tn) {
            tf.findReferencedTypes(tn)
        }
    }
}

func (tf *TransitiveTypeFinder) findReferencedTypes(root astmodel.TypeName) {
    tf.applicableTypes.Add(root)
    for referenced := range tf.allTypes[root].Type().References() {
        if !tf.applicableTypes.Contains(referenced) {
            tf.findReferencedTypes(referenced)
        }
    }
}
```

Then usage becomes:
```
case ExportFilterIncludeTransitive:
    typeFinder := &TransitiveTypeFinder{allTypes: allTypes, applicableTypes: make(astmodel.TypeNameSet), filter: f}
    typeFinder.Initialize()
    return typeFinder.ShouldExport // I think this works?
```

Reply:
I don't think this makes it much clearer, so I'll leave it as-is.